### ISSUE TITLE
Slow stream queries

### ISSUE THREAD

**Issue by amckinstry** [diaspora/diaspora#6523]
I run a small pod (a few dozen users, only). This has problems with very slow stream queries; the pod runs 0.5.4.0 now but is relatively old having being updated a few times, and so the likely issue is db-related rather than code.

I see the issue when browsing as a logged-in user; it can take minutes now for the stream to appear, making it unusable. However it works fine when running 'as a blog' with just accessing http://diaspora.sceal.ie/u/amckinstry without logging in, or when accessing with the D* app (which uses a different stream query?)

How do I debug such issues?

**Comment by amckinstry** [comment]
Notes: 
I originally ran a msql db, a few months ago I moved to psql which fixed some speed issues at the time (I had the problem back then too) but now its back.

**Comment by Flaburgan** [comment]
Please stop your pod, connect to the database with `psql diaspora_production` and run `vacuum full analyze verbose;`.

What do you talk about saying "diaspora app" by the way? The mobile version?

**Comment by amckinstry** [comment]
Thanks, did that but to no avail.

A number of removable rows were cleaned, but nothing unusual showed up.
The "diaspora app" is the Diaspora Native App for Android (from f-droid).

**Comment by amckinstry** [comment]
So, debugging I've checked the postgres logs.
I'm seeing errors like:

2015-10-25 07:44:25 GMT [21253-13] postgres@diaspora_production ERROR:  duplicate key value violates unique constraint "posts_guid_key"
2015-10-25 07:44:25 GMT [21253-14] postgres@diaspora_production DETAIL:  Key (guid)=(39aff3905b000133bec34860008dbc6c) already exists.
2015-10-25 07:44:25 GMT [21253-15] postgres@diaspora_production STATEMENT:  INSERT INTO "posts" ("type", "guid", "text", "author_id", "diaspora_handle", "public", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5, $6, $7, $8) RETURNING "id"
2015-10-25 07:44:25 GMT [1566-131] postgres@diaspora_production ERROR:  duplicate key value violates unique constraint "posts_guid_key"
2015-10-25 07:44:25 GMT [1566-132] postgres@diaspora_production DETAIL:  Key (guid)=(39aff3905b000133bec34860008dbc6c) already exists.
2015-10-25 07:44:25 GMT [1566-133] postgres@diaspora_production STATEMENT:  INSERT INTO "posts" ("type", "guid", "text", "author_id", "diaspora_handle", "public", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5, $6, $7, $8) RETURNING "id"
2015-10-25 08:02:54 GMT [21273-16] postgres@diaspora_production ERROR:  duplicate key value violates unique constraint "posts_guid_key"
2015-10-25 08:02:55 GMT [21273-17] postgres@diaspora_production DETAIL:  Key (guid)=(cbe2c7205d1c01331acf7e8f7b917a82) already exists.
2015-10-25 08:02:55 GMT [21273-18] postgres@diaspora_production STATEMENT:  INSERT INTO "posts" ("type", "guid", "text", "author_id", "diaspora_handle", "public", "created_at", "provider_display_name", "updated_at") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING "id"
2015-10-25 09:58:17 GMT [21284-10] postgres@diaspora_production ERROR:  duplicate key value violates unique constraint "posts_guid_key"
2015-10-25 09:58:17 GMT [21284-11] postgres@diaspora_production DETAIL:  Key (guid)=(f45a34e05d2c013331c900505600750e) already exists.
2015-10-25 09:58:17 GMT [21284-12] postgres@diaspora_production STATEMENT:  INSERT INTO "posts" ("type", "guid", "text", "author_id", "diaspora_handle", "public", "created_at", "provider_display_name", "updated_at") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING "id"
2015-10-25 10:28:22 GMT [21273-19] postgres@diaspora_production ERROR:  duplicate key value violates unique constraint "people_diaspora_handle_key"
2015-10-25 10:28:22 GMT [21273-20] postgres@diaspora_production DETAIL:  Key (diaspora_handle)=(slygoodvibes@framasphere.org) already exists.
2015-10-25 10:28:22 GMT [21273-21] postgres@diaspora_production STATEMENT:  INSERT INTO "people" ("diaspora_handle", "guid", "serialized_public_key", "url", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"
2015-10-25 10:39:58 GMT [1565-128] postgres@diaspora_production ERROR:  duplicate key value violates unique constraint "posts_guid_key"
2015-10-25 10:39:58 GMT [1565-129] postgres@diaspora_production DETAIL:  Key (guid)=(cb0d12a05d320133feb300259069449e) already exists.
2015-10-25 10:39:58 GMT [1565-130] postgres@diaspora_production STATEMENT:  INSERT INTO "posts" ("type", "guid", "text", "author_id", "diaspora_handle", "public", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5, $6, $7, $8) RETURNING "id"
2015-10-25 12:00:49 GMT [1566-134] postgres@diaspora_production ERROR:  cached plan must not change result type
2015-10-25 12:00:49 GMT [1566-135] postgres@diaspora_production STATEMENT:  SELECT  "posts".* FROM "posts" WHERE "posts"."type" IN ('Reshare') AND "posts"."author_id" = $1 AND "posts"."root_guid" = $2  ORDER BY "posts"."id" ASC LIMIT 1
2015-10-25 15:20:41 GMT [21273-22] postgres@diaspora_production ERROR:  duplicate key value violates unique constraint "posts_guid_key"

Any ideas how to fix these?

I've turned on logging slow queries (any > 100ms) via:
log_min_duration_statement = 100
#log_min_duration_statement = -1        # -1 is disabled, 0 logs all statements
                                        # and their durations, > 0 logs only
                                        # statements running at least this number
                                        # of milliseconds
in postgresql.conf

and now I see, eg.
2015-10-27 10:33:09 GMT [22432-6] postgres@diaspora_production LOG:  duration: 894.525 ms  execute <unnamed>: SELECT "tags".* FROM "tags" WHERE (LOWER(name) = LOWER('max') OR LOWER(name) = LOWER('igan') OR LOWER(name) = LOWER('attacked') OR LOWER(name) = LOWER('hospitalized') OR LOWER(name) = LOWER('australia') OR LOWER(name) = LOWER('nwo') OR LOWER(name) = LOWER('humanity') OR LOWER(name) = LOWER('insanity') OR LOWER(name) = LOWER('55') OR LOWER(name) = LOWER('cancer') OR LOWER(name) = LOWER('curable') OR LOWER(name) = LOWER('ian') OR LOWER(name) = LOWER('crane') OR LOWER(name) = LOWER('health') OR LOWER(name) = LOWER('chemo') OR LOWER(name) = LOWER('radiotherapy') OR LOWER(name) = LOWER('wtfu') OR LOWER(name) = LOWER('now'))
2015-10-27 10:33:25 GMT [32058-1] postgres@diaspora_production LOG:  duration: 199.306 ms  execute <unnamed>: SELECT "comments".* FROM "comments" WHERE "comments"."commentable_type" = 'Post' AND "comments"."commentable_id" IN (69083)
2015-10-27 10:34:15 GMT [32058-2] postgres@diaspora_production LOG:  duration: 270.154 ms  execute a1006: SELECT  "posts".* FROM "posts" WHERE "posts"."id" = $1  ORDER BY "posts"."id" ASC LIMIT 1
2015-10-27 10:34:15 GMT [32058-3] postgres@diaspora_production DETAIL:  parameters: $1 = '89914'
2015-10-27 10:34:41 GMT [32058-4] postgres@diaspora_production LOG:  duration: 703.474 ms  execute a1912: SELECT COUNT(*) FROM "people" INNER JOIN "profiles" ON "profiles"."person_id" = "people"."id" INNER JOIN "taggings" ON "taggings"."taggable_id" = "profiles"."id" AND "taggings"."context" = $1 AND "taggings"."taggable_type" = $2 INNER JOIN "tags" ON "tags"."id" = "taggings"."tag_id" WHERE "tags"."name" = 'mozilla' AND (profiles.searchable IS TRUE)
2015-10-27 10:34:41 GMT [32058-5] postgres@diaspora_production DETAIL:  parameters: $1 = 'tags', $2 = 'Profile'
2015-10-27 10:36:55 GMT [32058-6] postgres@diaspora_production LOG:  duration: 164.412 ms  execute a1918: SELECT   posts.* FROM "posts" WHERE "posts"."type" IN ('StatusMessage') AND "posts"."public" = $1 AND "posts"."pending" = $2 AND (EXISTS (SELECT 1 FROM taggings posts_taggings_3e06acb WHERE posts_taggings_3e06acb.taggable_id = "posts".id AND posts_taggings_3e06acb.taggable_type = 'Post' AND posts_taggings_3e06acb.tag_id in (32397))) AND (posts.created_at < '2015-10-27 10:36:56.000000') AND "posts"."type" IN ('StatusMessage', 'Reshare')  ORDER BY posts.created_at desc LIMIT 15
2015-10-27 10:36:55 GMT [32058-7] postgres@diaspora_production DETAIL:  parameters: $1 = 't', $2 = 'f'
2015-10-27 10:37:05 GMT [32058-8] postgres@diaspora_production LOG:  duration: 232.858 ms  execute a1919: SELECT   posts.* FROM "posts" WHERE "posts"."type" IN ('StatusMessage') AND "posts"."public" = $1 AND "posts"."pending" = $2 AND (EXISTS (SELECT 1 FROM taggings posts_taggings_c28ea2e WHERE posts_taggings_c28ea2e.taggable_id = "posts".id AND posts_taggings_c28ea2e.taggable_type = 'Post' AND posts_taggings_c28ea2e.tag_id in (19994))) AND (posts.created_at < '2015-10-27 10:37:05.000000') AND "posts"."type" IN ('StatusMessage', 'Reshare')  ORDER BY posts.created_at desc LIMIT 15
2015-10-27 10:37:05 GMT [32058-9] postgres@diaspora_production DETAIL:  parameters: $1 = 't', $2 = 'f'
2015-10-27 10:37:35 GMT [32058-10] postgres@diaspora_production LOG:  duration: 100.487 ms  statement: DEALLOCATE a920

**Comment by amckinstry** [comment]
I'm seeing a lot of:
2015-10-27 11:09:37 GMT [6027-7] postgres@diaspora_production LOG:  duration: 319.197 ms  execute <unnamed>: SELECT  1 AS one FROM "participations" WHERE "participations"."guid" = '2fb751405ec9013302f6425c375b904e' LIMIT 1
2015-10-27 11:09:37 GMT [6027-8] postgres@diaspora_production LOG:  duration: 312.522 ms  execute <unnamed>: SELECT  1 AS one FROM "participations" WHERE ("participations"."target_id" = 124220 AND "participations"."target_type" = 'Post' AND "participations"."author_id" = 369) LIMIT 1
2015-10-27 11:10:17 GMT [6028-10] postgres@diaspora_production LOG:  duration: 322.387 ms  execute a830: SELECT  1 AS one FROM "participations" WHERE "participations"."author_id" = $1 AND "participations"."target_id" = 124209 LIMIT 1

All around ~300ms. I notice that participations does not have any indexes. Might adding them help, and if so, how do I test that?

**Comment by jhass** [comment]
The failed constraints are harmless, ignore them.

I do have indexes on participations: ![](http://cloud.aeshna.de/u/mrzyx/screenshots/screenshot_20151027_124301.png)

**Comment by amckinstry** [comment]
So, my db schema got out of sync somehow despite (?) an upgrade to 0.5.4.0.
I had git checkout'd the repo, but my db/* contents were different. I pulled the latest, re-did the migrate.

I was still missing some indexes, which I wrote a script to insert by hand (my own db/migrate/*.rb script following the examples). Most of my performance issues have disappeared.

I'm back to 1-3 second stream updates, which I can blame on underpowered server at home.

Thanks
